### PR TITLE
MethodSpec factory for overriding an ExecutableElement.

### LINKED ISSUE #228: Factory for MethodSpec.Builder for overriding ExecutableElement
I'm working in a place where I need to override methods. It would be nice to get a method builder of the same name, return type, params, throws declarations, and already having `@Override`on it.

Maybe `MethodSpec.methodBuilderOverriding(ExecutableElement)`?

### PR DESCRIPTION
**Pull request by JakeWharton**
Anyone know a good way to test this without writing a novel of mocking? I filed https://github.com/google/compile-testing/issues/65 which seems like it would be a good approach.

Closes #228.

### CODE REVIEW

**Comment by JakeWharton** [comment]
Also, two questions:
 * Do we copy method annotations?
 * Do we copy param annotations?

Copying type annotations seems like an obvious yes but we don't support that yet (#136).

**Comment by swankjesse** [comment]
Make sure you don't get two `@Override` annotations if one is already present.

On copying over annotations, we should do what IntelliJ does: copy 'em. Though we need API to remove annotations too.

On testing, can you do what TypesTest does to get Elements for the current file?

**Comment by JakeWharton** [comment]
PTAL.

Copying annotations requires non-trivial infrastructure around being able to create `AnnotationSpec`'s from `AnnotationMirrors` which I'll do in a follow-up.

File: src/main/java/com/squareup/javapoet/MethodSpec.java [hunk 1]
```diff
@@ -150,6 +156,52 @@ public static Builder constructorBuilder() {
     return new Builder(CONSTRUCTOR);
   }
 
+  /**
+   * Create a builder which overrides the specified {@code method}. This will copy its visibility
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/MethodSpec.java]: nit: drop _the specified_. Using code font is good enough!

File: src/main/java/com/squareup/javapoet/MethodSpec.java [hunk 2]
```diff
@@ -150,6 +156,52 @@ public static Builder constructorBuilder() {
...
+    if (modifiers.contains(Modifier.PRIVATE)
+        || modifiers.contains(Modifier.FINAL)
+        || modifiers.contains(Modifier.STATIC)) {
+      throw new IllegalArgumentException("cannot override method with modifiers: " + modifiers);
+    }
+
+    String methodName = method.getSimpleName().toString();
+    MethodSpec.Builder methodBuilder = MethodSpec.methodBuilder(methodName);
+
+    // TODO copy method annotations.
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/MethodSpec.java]: add an `@Ignore` failing test for these?

File: src/main/java/com/squareup/javapoet/MethodSpec.java [hunk 3]
```diff
@@ -150,6 +156,52 @@ public static Builder constructorBuilder() {
...
+    // TODO check to ensure we're not duplicating override annotation.
+    methodBuilder.addAnnotation(Override.class);
+
+    modifiers = new LinkedHashSet<>(modifiers); // Local copy so we can remove.
+    modifiers.remove(Modifier.ABSTRACT);
+    methodBuilder.addModifiers(modifiers.toArray(new Modifier[modifiers.size()]));
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/MethodSpec.java]: We probably want an overload of `addModifiers()` that takes a `Collection` !

File: src/main/java/com/squareup/javapoet/MethodSpec.java [hunk 4]
```diff
@@ -150,6 +156,52 @@ public static Builder constructorBuilder() {
...
+
+    for (TypeParameterElement typeParameterElement : method.getTypeParameters()) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/MethodSpec.java]: Off topic: I can't think of an overriding method with type parameters.

File: src/test/java/com/squareup/javapoet/MethodSpecTest.java [hunk 5]
```diff
@@ -15,20 +15,39 @@
...
+  }
 
   @Test public void nullAnnotationsAddition() {
     try {
       MethodSpec.methodBuilder("doSomething").addAnnotations(null);
       fail();
     } catch (IllegalArgumentException expected) {
-      assertThat(expected.getMessage())
-          .isEqualTo("annotationSpecs == null");
+      assertThat(expected).hasMessage("annotationSpecs == null");
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/MethodSpecTest.java]: Nice.

File: src/test/java/com/squareup/javapoet/MethodSpecTest.java [hunk 6]
```diff
@@ -15,20 +15,39 @@
...
+import java.util.List;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.TypeElement;
+import org.junit.Rule;
 import org.junit.Test;
 
+import static com.google.common.collect.Iterables.getOnlyElement;
 import static com.google.common.truth.Truth.assertThat;
+import static javax.lang.model.util.ElementFilter.methodsIn;
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/MethodSpecTest.java]: neat.

File: src/test/java/com/squareup/javapoet/MethodSpecTest.java [hunk 7]
```diff
@@ -57,8 +74,55 @@
...
+
+  @Test public void overrideEverything() {
+    TypeElement classElement = getElement(Everything.class);
+    ExecutableElement methodElement = getOnlyElement(methodsIn(classElement.getEnclosedElements()));
+
+    MethodSpec method = MethodSpec.overriding(methodElement).build();
+    assertThat(method.toString()).isEqualTo(""
+        + "@java.lang.Override\n"
+        + "protected <T extends java.lang.Runnable & java.io.Closeable> "
+        + "java.lang.Runnable everything(java.lang.String arg0, java.util.List<? extends T> arg1) "
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/MethodSpecTest.java]: arg0, arg1 is very sad!

**Comment by swankjesse** [comment]
LGTM

File: src/test/java/com/squareup/javapoet/MethodSpecTest.java (hunk 7 above)
Comment (JakeWharton) [inline src/test/java/com/squareup/javapoet/MethodSpecTest.java]: ew, yeah. wtf. that shouldn't happen.

File: src/main/java/com/squareup/javapoet/MethodSpec.java (hunk 3 above)
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/MethodSpec.java]: ha, yeah. not sure why I didn't just add one. will do.

File: src/test/java/com/squareup/javapoet/MethodSpecTest.java (hunk 7 above)
Comment (JakeWharton) [inline src/test/java/com/squareup/javapoet/MethodSpecTest.java]: I'm pretty sure this is just due to our use of compile-testing which references back into the elements of the test code (cc/ @gk5885). In an annotation processor, you'd have the real names.

File: src/main/java/com/squareup/javapoet/MethodSpec.java (hunk 2 above)
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/MethodSpec.java]: done

**Pull request merged**